File: my_code/DSA/PYTHON/DS/stack and queue/stack_list_implementation_top.py
# linked list implementation [top]

class Node:
    def __init__(self, val):
        self.val = val
        self.next = None

class Stack:
    def __init__(self):
        self.top = None

    def peek(self):
        if not self.top:
            return None
        else:
            return self.top.val
        
    def is_empty(self):
        if not self.top:
            return True
        else:
            return False
    
    def clear(self):
        # remove the access (pointer which catch the list)
        self.top = None
    
    # add to the top of the list
    # its unshift but I called it push
    # to be O(1), while real push take O(n)
    def push(self, val):
        # create a new node to add it
        new_node = Node(val)
        # set the newNode next to point to the previous start
        new_node.next = self.top
        #  set the start to be the new node
        self.top = new_node

        # and return it
        return self
    
    # remove the top node of the list
    # its shift but I called it push
    # to be O(1), while real pop take O(n)
    def pop(self):
        if not self.top:
            return None
        
        # store the last in (top node) to return after removing
        popped_node = self.top
        # make the top node be the next node
        self.top = self.top.next

        return popped_node.val
    

# Commenting for Testing
# stack = Stack()
# stack.push(32)
# stack.push(12)
# stack.push(65)
# stack.push(10)
# print(stack.pop())
# print(stack.pop())
# print(stack.pop())
# print(stack.pop())

File: my_code/DSA/JS/DS/stack and queue/stack_list_implementation_top.js
// linked list implementation [top]

class Node {
  // each element in the stack must have data and pointer to the nest element
  constructor(data) {
    this.data = data;
    this.next = null;
  }
}

class Stack {
  constructor() {
    // top pointer for each stack instance to track the top element
    this.top = null;
  }

  // time complexity O(1)
  peek() {
    // if the top [last element] is not empty [None] return its data
    if (!this.top) {
      return null;
    } else {
      return this.top.data;
    }
  }

  // time complexity O(1)
  isEmpty() {
    // if the top is empty then return true
    if (!this.top) {
      return true;
    } else {
      return false;
    }
  }

  // time complexity O(1)
  clear() {
    // remove the access (pointer which catch the list)
    // so now my pointer not refer anywhere
    this.top = null;
  }

  // push the data to the start of the list [top]
  // because in list the top refer to the first element O(1)
  // but in our stack the top refer to the last in element
  // time complexity O(1)
  push(data) {
    // create a new node to add it
    const newNode = new Node(data);
    // set the newNode next to point to the previous start of the list
    newNode.next = this.top;
    // set the start to be the new node
    this.top = newNode;

    // and return it
    return this;
  }

  // remove the data to the start of the list [top]
  // because in list the top refer to the first element O(1)
  // but in our stack the top refer to the last in element
  // time complexity O(1)
  pop() {
    // if the stack is empty return null
    if (this.isEmpty()) {
      return null;
    }

    // store the last in (first node) to return after removing
    const poppedNode = this.top;
    // make the first node be the next node
    this.top = this.top.next;

    return poppedNode.data;
  }
}

module.exports = Stack;
